### ISSUE TITLE
[[CSL STYLE ERROR: reference with no printed form.]] with citeprocNoteCitekey TRUE

### ISSUE THREAD

**Issue by elliotmartin92** [retorquere/zotero-better-bibtex#1830]
**Support log ID:**
9UQVJKRI-euc
**Expected behavior:**
When using the citation style "Better BibTeX Citekeys" with citeprocNoteCitekey TRUE, citations should appear as citekeys
**Actual behavior:**
Citations appear as the following error: "[[CSL STYLE ERROR: reference with no printed form.]]".

This used to work for me as of ~6 months ago. I've also looked into previous questions regarding this error, which in those peoples cases was resolved by properly changing citeprocNoteCitekey, however, it is properly enabled here.

Thanks for any help,

**Comment by retorquere** [comment]
Can you get me a new log from build 990 after reproducing the problem?

**Comment by elliotmartin92** [comment]
Log after installing build 990: 6D6WLLLM-euc

**Comment by retorquere** [comment]
Does a refresh from the Zotero toolbar in word help? I don't see Zotero handing out new reference data, so it might be that the rendering process is re-using the reference data stored inside word.

**Comment by elliotmartin92** [comment]
No, refreshing doesn't help and in fact what I am doing prior to sending you the log is switching the citation style from Chicago to Better BibTeX Citekeys, so it should have refreshed already.

**Comment by retorquere** [comment]
Apparently that doesn't cause the Word plugin to re-fetch the CSL data, which is required to get the patched data. I'll ask on the Zotero forums.

**Comment by retorquere** [comment]
A new build will drop here shortly -- can you get me a new log?

**Comment by elliotmartin92** [comment]
Log after installing build 991 Z3KJYMX3-euc

**Comment by retorquere** [comment]
We'll have to wait for [this](https://groups.google.com/g/zotero-dev/c/TJYRyktcaDg).

**Comment by elliotmartin92** [comment]
Sounds good, perhaps a silly question though, if citeprocNoteCitekey==TRUE, should I be able to see the citekey in the "Notes" tab/field in Zotero? Because most of the notes fields are blank? Or are you saying that the note field is only filled with the citekey upon update and that's the problem?

**Comment by retorquere** [comment]
Not silly, this is all pretty arcane; Zotero converts its items to CSL-JSON format before sending it to word, and uses that CSL-JSON format to create the citations/bibliography. CSL knows about a limited number of fields, but one of the supported fields is `note`, so I apply a run-time-patch to zotero when citeprocNoteCitekey==TRUE to replace the contents of the `note` field with the citekey on the fly as it creates the CSL-JSON from the item, but that requires Zotero to actually call that method. Currently it does not, and I don't know how to make it do so - that's the question I posted on the forums.

This onl-the-fly transformation only happens to the transient CSL-JSON representation of the item, it won't show up anywhere else.

**Comment by retorquere** [comment]
I'm not getting any traction on zotero-dev it seems. Maybe you can ask on the regular Zotero forums, the devs seem more active there.

**Comment by elliotmartin92** [comment]
I've poster there, I'll update when the post is approved.

**Comment by elliotmartin92** [comment]
It appears as though my post to the forums never got approved. I've resubmitted it and will update if it gets approval.

**Comment by retorquere** [comment]
I didn't know https://forums.zotero.org/ required approval?

**Comment by elliotmartin92** [comment]
When I submit the post post a pop-up appears saying that my post will go up pending mod approval. I've messaged "dstillman", who seems to be an active mod there to try and get my post approved. I suspect I am getting caught by a spam filter as I included a link here and to your post on the dev-group to give context.

**Comment by elliotmartin92** [comment]
Ya. I am totally unable to post on the Zotero forums. I tried one last time, this time removing any leaks any copied the exact message I receive: "Posts from new users are moderated. Your message will appear after it has been approved". My message to a mod hasn't been replied to though they have been active since I sent it.

**Comment by retorquere** [comment]
What is your username on the forums?

**Comment by elliotmartin92** [comment]
etm4
________________________________

**Comment by retorquere** [comment]
Can you clear BBTs caches and try again? The cache ought to have dropped when you enabled citeprocnotecitekey, but if it wasn't that would in fact explain why you're not seeing a refresh - it would be served from a stale cache.

**Comment by elliotmartin92** [comment]
The zotero forum post was approved, apparently it wasn't getting approved as you had posted the question in the dev forum.

https://forums.zotero.org/discussion/90337/csl-data-in-a-word-document-not-refreshing#latest

**Comment by retorquere** [comment]
Can you start with `citeprocNoteCitekey` on, switch styles as indicated on the forum, and send a new debug log?

**Comment by elliotmartin92** [comment]
Attempted that. The result is the same which I think is unsurprising, after all I was already switching styles once. Here I just toggled it back and forth.

Log ID: 447VSPB7-euc

**Comment by retorquere** [comment]
Can you try again with 1105?

**Comment by elliotmartin92** [comment]
Tried again: JML7H5T6-euc

**Comment by elliotmartin92** [comment]
I have found something odd. I attempted to use Better BibTeX Citekeys on a different document on the same machine and it worked. 

The only reason I can think that this would be the case is that the document that is not working was originally created using Mendeley while the document that works may have been done with Zotero from the beginning (I can't remember when I switched over). 

However, besides this one strange bug the document that doesn't work seems to be perfectly linked to my Zotero database. I can use any other citation style, I can add/delete/view the citations, etc.

I hope I haven't made a silly blunder sending you down this rabbit hole.

**Comment by retorquere** [comment]
Can you start with debugging enabled from the help menu, replicate the problem, and send a new log? I still see zero activity and that would indicate it's just not enabled for you.

I don't know what to make of the ex-Mendeley document. If you can change it between styles and it works (and you don't have Mendeley running) that means it's connected to Zotero. But I thought that there was no automated migration for documents from Mendeley (or anything, really) to Zotero markers.

**Comment by elliotmartin92** [comment]
Debugging should have been enabled. But I've ensured that it is and done the following. Change style from Chicago->Better BibTeX->ACS->Better BibTeX and then submitted a log.

Log ID: FT8XH5LL-euc

**Comment by retorquere** [comment]
The problem is that the log shows no indication of even trying to install the monkey-patch. Can you try again with the build that will drop here in a few minutes?

**Comment by elliotmartin92** [comment]
Log with patch 5.4.25.1110: 4G3XQY3P-euc

**Comment by retorquere** [comment]
Yeah OK, I can see it's enabled properly now, but it's just not being called. 

> As for migration, I followed the guide here: https://www.zotero.org/support/kb/mendeley_import.

That is for the database content though. This does nothing with documents that were previously linked to Mendeley. Citations in Word documents that previously worked with Mendeley must, as far as I know, be manually re-entered to get the document connected to Zotero. I don't know of any automated procedure to change the mendeley-connected citation markers in a word document to zotero-connected citation markers, so for a document where you  have not manually re-entered these I would expect the document to still have mendeley-connected markers. But if that is true, and Mendeley is not running, and you don't have the Mendeley word-plugin installed, and you can switch that document between eg APA and (blergh) IEEE, that means it must be connected to Zotero.

Can you confirm the scenario above applies to you?

**Comment by elliotmartin92** [comment]
So I can confirm that I can change the citation style through Zotero. However, if I right click the citation the `update field` item is greyed out. This is contrast to my document that was done with Zotero from the beginning. It seems that although I can change the style, the citations are not actually linked to Zotero and it also seems the only option is to manually recite the the document. 

This behavior is alluded to here: https://forums.zotero.org/discussion/74608/citations-not-updating-on-refresh

This wasn't an issue with BetterBibTeX at all so I'll close the issue unless you have any other thoughts.

**Comment by retorquere** [comment]
Interesting -- so the Zotero Word plugin apparently sends the embedded CSL-JSON from Word to Zotero to render bibliographies for unconnected documents. That's nifty, but of no help here. Thanks for the diagnosis!

**Comment by elliotmartin92** [comment]
So I'd like to add some context and describe how I ultimately resolved this issue in case anyone stumbles on this thread with similar problems as I was having. Sorry if this is not really related to better-bibtex or the original issue I opened, but better-bibtex does play an important role here in allowing you to see/edit citekeys.

My goal is to convert several documents that have been written over the past 6 years during my PhD in order to construct my thesis.
1. I want to avoid having to "recite" each document. 
2. I want to convert the documents to Rmarkdown documents to use them in thesisdown

My resolution was as follows:
I had to "unify" the citekeys used by Mendeley and Zotero, so I could export a Bibtex from Zotero that I can use with my markdown document. If I didn't do this I would have duplicate entries in my bibliography.

To do this one must install a version of Mendeley from before they encrypted your citation database. Then borrowing the approach from `better-bibtex` , edit the SQL database to copy the citation-keys to an unused field (such as short-title). Edit the better-bibtex CSL to use the field you picked and use it to make markdown style citations. 

Next, import your Mendeley library into Zotero using Zotero's import function. Resolve any duplicates keeping the Mendeley version to preserve the Mendeley citekey. Now the databases are "unified".

Refresh any documents that used Zotero for citations to update the citekeys to the unified ones and perform document conversion.
